Fix: stonith-history: keep history recorded within boundaries

wenningerk opened the pull request:
for discussion ...

introduced stonith_fence_history_trim so that algorithm used there can easily be switched from currently purging of the whole history to e.g. purging of just half of the history or whatever seems convenient.
MAX_STONITH_HISTORY is available globally so that the decision when to trigger the trim (and if it should be cluster-wide or just locally - e.g. if you want it done synchronously before proceeding) is cheap.

kgaillot commented on daemons/fenced/fenced_history.c:
```diff
@@ -98,6 +99,47 @@ stonith_fence_history_cleanup(xmlNode *request, const char *target)
+stonith_fence_history_trim(gboolean broadcast)
+{
+    crm_warn("Fencing History growing beyond limit of %d so purge "
+             "failed/successful attempts.", MAX_STONITH_HISTORY);
```
No "." needed at end of log messages

kgaillot commented on daemons/fenced/fenced_history.c:
```diff
@@ -98,6 +99,47 @@ stonith_fence_history_cleanup(xmlNode *request, const char *target)
+/* keeping the length of fence-history within bounds
+ * =================================================
+ *
+ * If things are really running wild a lot of fecing-atempts
```
spelling

kgaillot commented on daemons/fenced/fenced_history.c:
```diff
@@ -240,6 +282,11 @@ stonith_merge_in_history_list(GHashTable *history)
             continue; /* skip over existant - state-merging migh be desirable */
         }
 
+        if (g_hash_table_size(stonith_remote_op_list) >= MAX_STONITH_HISTORY) {
```
Rather than count the table every time, it would be more efficient to count it once at the beginning, and increment the count for each merged op. Or even wait until after the loop to check it once, though that means we might briefly go over the limit.

wenningerk replied:
That was my initial approach as well but looking at the implementation I realized that the counter is already part of the table so that it is cheap and thus not worth risking my private counter to get out of sync.

kgaillot replied:
Ah, I didn't realize GHashTable kept a counter. Guess I was thinking of GList

wenningerk replied:
Considering a smarter trim it probably makes sense to go over the limit and trim then. Anyway there is no significant additional memory-consumption as it is just moving ops between two hash-tables.

kgaillot commented on daemons/fenced/fenced_history.c:
```diff
@@ -240,6 +282,11 @@ stonith_merge_in_history_list(GHashTable *history)
         }
 
+        if (g_hash_table_size(stonith_remote_op_list) >= MAX_STONITH_HISTORY) {
+            stonith_fence_history_trim(TRUE);
```
If all the nodes merge after a partition, and it brings them over the limit, I'm a little concerned about the redundancy of every node requesting a trim. But I'm not sure it can be avoided.

kgaillot commented on daemons/fenced/pacemaker-fenced.h:
```diff
@@ -7,6 +7,8 @@
 
 #include <crm/common/mainloop.h>
 
+#define MAX_STONITH_HISTORY 100
```
I'm not too concerned about the exact number here, because the logs serve as an alternate record, and because if stonith is having a high number of repeated failures, they are likely identical, and the user will likely need to look at the logs to figure out what is going on anyway.

I'm guessing each record averages around 60 bytes, so this would take up about 6K. I think we'd be fine with anything from 100-500.

wenningerk replied:
With the xml-overhead + possibly long node-names it might be a few 100 bytes per entry but as we have cib-messages >100K that still shouldn't matter.

wenningerk commented:
switched to just local trimming of the history to half of a maximum size once over the maximum.
Diffs from other nodes are always merged in first and we have a trim afterwards so
that we have the correct entries kicked out. (memory is uncritical as the entries are anyway
already allocated and just moved from one hash to another)
The trim-function does the checking for the maximum-size internally with a little performance
penalty for the function-call in exchange for a more local implementation.

kgaillot commented on daemons/fenced/fenced_history.c:
```diff
@@ -98,6 +101,109 @@ stonith_fence_history_cleanup(xmlNode *request, const char *target)
+                  "half of failed/successful attempts", MAX_STONITH_HISTORY);
+
+        /* write all ops into an array */
+        for (i=0, g_hash_table_iter_init(&iter, stonith_remote_op_list);
```
I believe some static analyzers will complain about a void value in a comma statement -- better to put it separate before.

wenningerk replied:
I try to avoid 'for' with parts of the initialization before and parts contained.
So let's go with the common 'while' on hash-table iteration ...

kgaillot commented on daemons/fenced/fenced_history.c:
```diff
@@ -98,6 +101,109 @@ stonith_fence_history_cleanup(xmlNode *request, const char *target)
+ * Simplest thing though is to purge the full history
+ * throughout the cluster once MAX_STONITH_HISTORY is reached.
+ * On the other hand this leads to purging the history in
+ * situations where it would be handy to have it probably.
```
Last few sentences are no longer needed, right?

wenningerk replied:
Wanted to document all ideas and why I wouldn't go for them atm. So I added the last sentence to state why I didn't go for the obviously simplest approach.
Same in the merge-code where I stated in a comment at the point where one would do a complete purge why I finish the merge first and do the trim afterwards.

kgaillot commented on daemons/fenced/fenced_history.c:
```diff
@@ -224,12 +329,14 @@ stonith_merge_in_history_list(GHashTable *history)
     GHashTableIter iter;
     remote_fencing_op_t *op = NULL;
     gboolean updated = FALSE;
+    guint op_list_size;
```
Not needed

wenningerk replied:
left over from checking the length outside of the trim-call.
actually interesting that the compiler didn't moan.
